server/src/com/cloud/storage/StorageManagerImpl.java

```diff
@@ -2199,15 +2199,20 @@ public void cleanupDownloadUrls(){
                 if(downloadUrlCurrentAgeInSecs < _downloadUrlExpirationInterval){  // URL hasnt expired yet
                     continue;
                 }
-
-                s_logger.debug("Removing download url " + volumeOnImageStore.getExtractUrl() + " for volume id " + volumeOnImageStore.getVolumeId());
+                long volumeId = volumeOnImageStore.getVolumeId();
+                s_logger.debug("Removing download url " + volumeOnImageStore.getExtractUrl() + " for volume id " + volumeId);
 
                 // Remove it from image store
                 ImageStoreEntity secStore = (ImageStoreEntity) _dataStoreMgr.getDataStore(volumeOnImageStore.getDataStoreId(), DataStoreRole.Image);
                 secStore.deleteExtractUrl(volumeOnImageStore.getInstallPath(), volumeOnImageStore.getExtractUrl(), Upload.Type.VOLUME);
 
                 // Now expunge it from DB since this entry was created only for download purpose
                 _volumeStoreDao.expunge(volumeOnImageStore.getId());
+                Volume volume = _volumeDao.findById(volumeId);
+                if (volume != null && volume.getState() == Volume.State.Expunged)
+                {
+                    _volumeDao.remove(volumeId);
+                }
```

Comment:
There might be multiple volume snapshot on image store.
We need to make sure that all volume snapshot have been removed from image store.

```
+                VolumeDataStoreVO volumesOnImageStore = _volumeStoreDao.findByVolume(volumeId);
+                if (volumesOnImageStore == null) { // volume is expunged from primary, as well as on secondary
+                    VolumeVO vol = _volumeDao.findById(volumeId);
+                    if (vol != null && (vol.getState() == Volume.State.Expunged)) {
+                        _volumeDao.remove(volumeId);
+                    }
+                }
```